Sources/CATransform3D+interpolate.swift

```diff
@@ -0,0 +1,237 @@
+        return transform
+    }
+
+    static func slerp(from q1: Quaternion, to q2: Quaternion, progress: Float) -> Quaternion {
+        // Compute the cosine of the angle between the two quaternions
+        var cosTheta = q1.x * q2.x + q1.y * q2.y + q1.z * q2.z + q1.w * q2.w
+
+        // If cosTheta < 0, the interpolation will take the long way around the sphere.
+        // To fix this, one quaternion must be negated.
+        var q2 = q2
+        if cosTheta < 0.0 {
+            q2 = Quaternion(x: -q2.x, y: -q2.y, z: -q2.z, w: -q2.w)
+            cosTheta = -cosTheta
+        }
+
+        // If the quaternions are too close, use linear interpolation
+        if cosTheta > 0.9995 {
+            let result = Quaternion(
+                x: q1.x + progress * (q2.x - q1.x),
+                y: q1.y + progress * (q2.y - q1.y),
+                z: q1.z + progress * (q2.z - q1.z),
+                w: q1.w + progress * (q2.w - q1.w)
+            )
+            return normalize(result)
+        }
+
+        // Perform the slerp interpolation
+        let angle = acos(cosTheta)
+        let sinTheta = sqrt(1.0 - cosTheta * cosTheta)
+        let a = sin((1.0 - progress) * angle) / sinTheta
+        let b = sin(progress * angle) / sinTheta
+
+        return Quaternion(
+            x: a * q1.x + b * q2.x,
+            y: a * q1.y + b * q2.y,
+            z: a * q1.z + b * q2.z,
+            w: a * q1.w + b * q2.w
+        )
+    }
+
+    static func normalize(_ q: Quaternion) -> Quaternion {
+        let magnitude = sqrt(q.x * q.x + q.y * q.y + q.z * q.z + q.w * q.w)
+        return Quaternion(x: q.x / magnitude, y: q.y / magnitude, z: q.z / magnitude, w: q.w / magnitude)
+    }
+}
+
+private extension DecomposedTransform {
+    func recompose() -> CATransform3D {
+        let transform = self
+
+        // Create scale matrix
+        var scaleMatrix = CATransform3DIdentity
+        scaleMatrix.m11 = transform.scale.x
+        scaleMatrix.m22 = transform.scale.y
+        scaleMatrix.m33 = transform.scale.z
+
+        // Create translation matrix
+        var translationMatrix = CATransform3DIdentity
+        translationMatrix.m41 = transform.translation.x
+        translationMatrix.m42 = transform.translation.y
+        translationMatrix.m43 = transform.translation.z
+
+        // Convert quaternion to rotation matrix
+        let rotationMatrix = transform.rotation.toTransform()
+
+        // Combine all matrices
+        var resultTransform = CATransform3DIdentity
+        resultTransform = CATransform3DConcat(scaleMatrix, rotationMatrix)
+        resultTransform = CATransform3DConcat(resultTransform, translationMatrix)
+
+        return resultTransform
+    }
+}
+
+#if os(Android)
+private func sqrt(_ value: Float) -> Float {
+    return Float(sqrt(Double(value)))
+}
+private func acos(_ value: Float) -> Float {
+    return Float(acos(Double(value)))
+}
+private func sin(_ value: Float) -> Float {
+    return Float(sin(Double(value)))
+}
+#endif
```

Comment:
These functions expect doubles on android. Is there a better workaround than this? The type conversions might cause some minor overhead

Reply:
aren't there float equivalents `sqrtf`, `acosf` and `sinf`?

Reply:
Nice thanks, I didn't know these exist. I cleaned this up in https://github.com/flowkey/UIKit-cross-platform/pull/399/commits/ced4e407a48f91d1992f0207b632e40d7e1a2e66

Reply:
the float equivalents are C functions that are available in `import Glibc` (which is actually the "Bionic" library and not Glibc – Android is a mess). They are also available on Apple platforms if you `import Darwin`.

For squareRoot specifically I would recommend against using free-standing functions and use the versions from the Swift standard library (`myFloat.squareRoot()`). The trigonometry functions don't exist in that form unfortunately so still need to be imported from somewhere

Reply:
The long term solution to this is to use `import Numerics`, which is available as a Swift package. It also introduces more performance and correctness fixes (including for trig functions)

https://www.swift.org/blog/numerics/

Reply:
cool, thanks for the feedback. I changed the lines where I used `sqrtf()` to use `. squareRoot()`. I'll look into using the Numerics package in a separate iteration. It looks interesting, thanks for the hint!